src/fnft_kdvv.c

```diff
@@ -1,209 +1,967 @@
+    }
+
+    if (bound_squared==0.0){
+        // Non-positive potentials have no bound states
+        *K_ptr=0;
+        goto leave_fun;
+    }
+
+    REAL const bounding_box[4] = {0,0,0,SQRT(bound_squared)};
+
+    // Localize bound states ...
+    switch (opts->bound_state_localization) {
+        case kdvv_bsloc_GRIDSEARCH_AND_REFINE:
+        {
+            // Set the search grid
+            UINT const M = 1000;
+            COMPLEX const eps_xi = I*(bounding_box[3]-bounding_box[2])/(M - 1);
+            xi = malloc(M * sizeof(COMPLEX));
+            if (xi == NULL) {
+                ret_code = E_NOMEM;
+                goto leave_fun;
+            }
+
+            xi[0] = I*nexttoward(bounding_box[2],bounding_box[3]);
+            for (i = 1; i < M-1; i++)
+                xi[i] = I * CIMAG(eps_xi*i);
+            xi[M-1] = I*nexttoward(bounding_box[3],bounding_box[2]);
+
+            // Allocate memory for call to kdv_scatter_matrix
+            scatter_coeffs = malloc(4 * M * sizeof(COMPLEX));
+            if (scatter_coeffs == NULL) {
+                ret_code = E_NOMEM;
+                goto leave_fun;
+            }
+
+            // Compute a(xi) on the grid on the imaginary axis
+            UINT const derivative_flag = 0;
+            ret_code = kdv_scatter_matrix(D, q, r, eps_t, kappa,
+                                          M, xi, scatter_coeffs ,opts_slow.discretization,derivative_flag);
+            CHECK_RETCODE(ret_code, leave_fun);
 
-release_mem:
-    free(transfer_matrix);
+            // Search for zerocrossings of a(xi)
+            K=0;
+            for (i = 0; i < M; i++) {
+                if ( scatter_coeffs[4*i]==0) {
+                    if (K>=*K_ptr) { // Lucky us: Exact bound state found
+                        ret_code = E_OTHER("More than *K_ptr initial guesses for bound states found. Increase *K_ptr and try again.")
+                        CHECK_RETCODE(ret_code, leave_fun);
+                    }
+                    bound_states[K]=xi[i];
+                    K++;
+                } else if ( i>0 && CREAL(scatter_coeffs[4*i])*CREAL(scatter_coeffs[4*(i-1)])<0) {
+                    if (K>=*K_ptr) { // Bound state between this sample and the previous
+                        ret_code = E_OTHER("More than *K_ptr initial guesses for bound states found. Increase *K_ptr and try again.")
+                        CHECK_RETCODE(ret_code, leave_fun);
+                    }
+                    // Zero crossing estimate with regula falsi:
+                    bound_states[K] = (xi[i-1] * CREAL(scatter_coeffs[4*i]) - xi[i] * CREAL(scatter_coeffs[4*(i-1)])) / CREAL( scatter_coeffs[4*i] - scatter_coeffs[4*(i-1)]);
+                    K++;
+                }
+            }
+        }
+            // fall through
+        case kdvv_bsloc_NEWTON: // Refine using Newton's method
+
+            // Perform Newton iterations. Initial guesses of bound-states
+            // should be in the continuous-time domain.
+
```

Comment:
Should there be a check on the validity of the initial guesses here?

Reply:
Fixed in https://github.com/FastNFT/FNFT/pull/64/commits/88aa44d4d89786db0c027409d72f12c238e4a4ba , https://github.com/FastNFT/FNFT/pull/64/commits/80fb3f67f51eb284e73ca3ecdecc9a90b2b822df